README
April 2004  (see end of file for copyright information)


This dicom to 4dfp conversion program is modified from the fine work
of the Electronic Radiology Lab, Washington University School of
Medicine, and Stephen M. Moore. Please respect their copyright
notices. At this time it should be used with caution. Further testing
is needed. The basic program was a dicom to 16 bit analyze image
conversion. Functions for producing 4dfp volume data have been
added. A rec file and ifh file are written.

The algorithm uses linked lists. A list of filenames is created. Then
files are opened and divided into analyze volumes using the series
time (dicom tag 0008 0031). This decision may be overriden by the -d
option. Each volume is first sorted by the relative image number (tag
0020 0013). Orientation and a check for multivolume data is
done. Default orientation is transverse. The volume is sorted again by
image position (0020 0032) to verify compliance with the 4dfp-Analyze
Coordinant System. If every image has the same position (bold), volume
size is set to 1.

Distance between slices (dime.pixdim[3]) is determined
preferentially. The default is to use the slice spacing tag 0018
0088. The slice thickness 0018 0050 will then be used. User defined
values for slice spacing may be set using the -t switch. (e.g. -t 2.5
would override the slice thickness and spaceing between slices tags,
and place the value 2.5 in the hdr file and the "scaling factor
(mm/pixel) [3]." in the ifh header.) The -c option may be used to
calculate this value using relative image position tag 0020 0032 and
will override any -t options or defaults. If tags for slice thickness
and slice spacing do not exist, and -t option is not used, the program
default will use the relative image position to calculate the distance
between slices.

Pixel values may be signed or unsigned 16 bit integers.  Min and max
are found checking each pixel.

Command Line Options:

-b  Output base filename follows the -b. The default is - analyze - 
-d  Divide series by the header tag given by the group and element number.
    ** Default will divide volumes using ID series time (0008 0031).
-f  Directories will be created, and dicom files will be moved.
-g  Add XYZ slice position, image name and number to rec file.

-c		Slice Spacing: Compute space between slices. (ignore -t)
-q		Slice Spacing: Do not compute space between slices (** Default)
-t <flt>	Slice Spacing: Use input value [-t <flt>]
-t T		Slice Spacing: Use Slice Thickness 0018 0050.[-t T]
-t S		Slice Spacing: Use Slice Spacing 0018 0088 [-t S](** Default)

-u  Output files named using sequence tag 0018 0024 plus number.

Analyze Coordinant System is determined using Image Position (0020 0032).
Multivolume and BOLD images are ordered by REL Image Number (0020 0013).

Reverse The Analyze Coordinate System Using Relative Image Position:
(Multivolume and BOLD data will reverse image numbering (0020 0013).)
-X  X Sagittal image positions will be ordered high to low.
-Y  Y Coronal image positions will be low to high.
-Z  Z Transverse image positions will be high to low.


COPYRIGHT INFO (August 2006)

The GNU Public License (see the file COPYING) applies only to GNU
software derivatives included in this package, notably "configure" and
arguably "config.h".  It does not apply to other files in this
directory; if there's a copyright notice at the top of the file,
believe what it says.


NEWS
30 August 2006  (karchie@npg.wustl.edu)

What are all these extra files (including, for example, NEWS)?  Most
are support for the GNU autotools, which is a (reasonably) portable
package for automatically generating Makefiles and handling various
system dependencies.

See INSTALL for how to use them.

If you got these files directly from CVS, you'll need to do a little
bit of setup before you can run "configure".  Do the following:

aclocal
autoheader
automake -a
autoconf

and then you can follow the instructions in INSTALL.

test_dcm.c
/*
 * $Id: test_dcm.c,v 1.1 2006/09/14 16:23:05 karchie Exp $
 * CUnit unit tests for dcm.c
 * by Kevin A. Archie, karchie@npg.wustl.edu
 */

#include <stdio.h>
#include <CUnit/Basic.h>

#include "dicom.h"
#include "condition.h"
#include "lst.h"
#include "dicom_objects.h"
#include "dicom_uids.h"
#include "dcmprivate.h"

extern CONDITION updateSpecialElements(PRIVATE_OBJECT **object,
				       PRV_ELEMENT_ITEM *item);
extern CONDITION newElementItem(DCM_ELEMENT *src, CTNBOOLEAN allocateData,
				PRV_ELEMENT_ITEM **dst);


/* The suite initialization function.
 * Returns zero on success, non-zero otherwise.
 */
int init_suite1(void)
{
  return 0;
}

/* The suite cleanup function.
 * Returns zero on success, non-zero otherwise.
 */
int clean_suite1(void)
{
  return 0;
}

#define BUF_SIZE 512

void test_updateSpecialElements(void) {
  PRIVATE_OBJECT *obj = 0;
  DCM_ELEMENT e;
  PRV_ELEMENT_ITEM *item = 0;
  DCM_TAG tag = DCM_METATRANSFERSYNTAX;
  DCM_VALUEREPRESENTATION vr = DCM_UI;
  CONDITION cond;
  char buf[BUF_SIZE];

  if (DCM_CreateObject((void*)&obj, 0) != DCM_NORMAL || obj == 0)
    CU_FAIL("Unable to create DCM object");
  

  memset(&e, 0, sizeof(e));
  e.tag = tag;
  e.representation = vr;
  e.d.string = buf;
  e.length = sizeof(buf) - 1;

  strcpy(buf, DICOM_TRANSFERLITTLEENDIAN);
  if (newElementItem(&e, FALSE, &item) != DCM_NORMAL || item == 0)
    CU_FAIL("Unable to create DCM Element Item");
  item->element.d.string = buf;
  cond = updateSpecialElements(&obj, item);
  CU_ASSERT_EQUAL(DCM_NORMAL, cond);
  CU_ASSERT_EQUAL(DCM_ORDERLITTLEENDIAN, obj->dataOptions);
  
  strcpy(buf, DICOM_TRANSFERLITTLEENDIANEXPLICIT);
  cond = updateSpecialElements(&obj, item);
  CU_ASSERT_EQUAL(DCM_NORMAL, cond);
  CU_ASSERT_EQUAL(DCM_EXPLICITLITTLEENDIAN, obj->dataOptions);

  strcpy(buf, DICOM_TRANSFERBIGENDIANEXPLICIT);
  cond = updateSpecialElements(&obj, item);
  CU_ASSERT_EQUAL(DCM_NORMAL, cond);
  CU_ASSERT_EQUAL(DCM_EXPLICITBIGENDIAN, obj->dataOptions);

  /* this one will generate warnings that don't interest us. */
  strcpy(buf, DICOM_TRANSFERJPEGEXTENDEDPROC2AND4);
  fprintf(stderr, "\nIgnore the following warning about transfer syntax:\n");
  cond = updateSpecialElements(&obj, item);
  CU_ASSERT_EQUAL(DCM_NORMAL, cond);
  CU_ASSERT_EQUAL(DCM_EXPLICITLITTLEENDIAN, obj->dataOptions);
}

void test_checkOrientation(void) {
  CU_FAIL("test unimplemented");
}


/* The main() function for setting up and running the tests.
 * Returns a CUE_SUCCESS on successful running, another
 * CUnit error code on failure.
 */
int main()
{
   CU_pSuite pSuite = NULL;

   /* initialize the CUnit test registry */
   if (CUE_SUCCESS != CU_initialize_registry())
      return CU_get_error();

   /* add a suite to the registry */
   pSuite = CU_add_suite("Suite_1", init_suite1, clean_suite1);
   if (NULL == pSuite) {
      CU_cleanup_registry();
      return CU_get_error();
   }

   /* add the tests to the suite */
   if (!CU_add_test(pSuite, "updateSpecialElements()", test_updateSpecialElements))
     {
       CU_cleanup_registry();
       return CU_get_error();
     }

   /* Run all tests using the CUnit Basic interface */
   CU_basic_set_mode(CU_BRM_VERBOSE);
   CU_basic_run_tests();
   CU_cleanup_registry();
   return CU_get_error();
}


test_dcm_to_4dfp.c
/*
 * $Id: test_dcm_to_4dfp.c,v 1.5 2006/11/09 17:19:38 karchie Exp $
 * CUnit unit tests for dcm_to_4dfp.c
 * by Kevin A. Archie, karchie@npg.wustl.edu
 */

#include <stdio.h>
#include <CUnit/Basic.h>

#include "dicom.h"
#include "condition.h"
#include "lst.h"
#include "dicom_objects.h"
#include "ifh.h"
#include "dcm_to_analyze.h"


/* The suite initialization function.
 * Returns zero on success, non-zero otherwise.
 */
int init_suite1(void)
{
  return 0;
}

/* The suite cleanup function.
 * Returns zero on success, non-zero otherwise.
 */
int clean_suite1(void)
{
  return 0;
}

#define BUF_SIZE 512

void test_setBase(void) {
  LST_HEAD *lh;
  PARAMS params;
  char buf[BUF_SIZE];

  strcpy(params.acqSequenceName, "*tse2dl_7");
  params.seriesNumber = 1;

  /* verify it works at all */
  lh = LST_Create();
  LST_Push(&lh, &params);
  setBase(&lh, buf);
  CU_ASSERT_STRING_EQUAL(buf, "tse2d_1");

  /* verify it works without leading *
   * and with nonzero chars already in the output buffer 
   */
  strcpy(params.acqSequenceName, "foobar328");
  params.seriesNumber = 3;
  setBase(&lh, buf);
  CU_ASSERT_STRING_EQUAL(buf, "fooba_3");

  /* how about short names? */
  memset(buf, 0, BUF_SIZE);
  strcpy(params.acqSequenceName, "*l");
  params.seriesNumber = 12;
  setBase(&lh, buf);
  CU_ASSERT_STRING_EQUAL(buf, "l_12");
}

void
test_testSelectionText(void) {
  DCM_OBJECT *obj = 0;
  CONDITION rval;
  DCM_ELEMENT e;
  DCM_TAG tag1 = DCM_IDREFERRINGPHYSICIAN; /* VR type Person's Name */
  DCM_TAG tag2 = DCM_IDOPERATORNAME; /* also VR type PN */
  char buf[1024];
  const char *testString = "Test Value";

  if (DCM_CreateObject(&obj, 0) != DCM_NORMAL || obj == 0)
    CU_FAIL("Unable to create DCM object");

  /* we reuse this element; okay because AddElement makes deep copy */
  memset(&e, 0, sizeof(e));
  e.tag = tag1;
  e.representation = DCM_PN;
  e.d.string = buf;
  e.length = sizeof(buf) - 1;

  if (DCM_AddElement(&obj, &e) != DCM_NORMAL)
    CU_FAIL("Unable to add element 1 to DCM object");
  rval = testSelectionText(&obj, tag1, testString);
  CU_ASSERT_EQUAL(rval, 1);

  e.tag = tag2;
  strcpy(buf, testString);
  if (DCM_AddElement(&obj, &e) != DCM_NORMAL)
    CU_FAIL("Unable to add element 2 to DCM object");
  rval = testSelectionText(&obj, tag1, testString);
  CU_ASSERT_EQUAL(rval, 1);	/* (0000,0001) still has empty text */
  rval = testSelectionText(&obj, tag2, testString);
  CU_ASSERT_EQUAL(rval, 0);	/* (0000,0002) has our text */
}



void
test_checkOrientation(void) {
  PARAMS paramsTra;
  PARAMS paramsCor;
  PARAMS paramsSag;
  PARAMS dummy1;
  PARAMS dummy2;
  float orientTra[] = { 1.0, 0.0, 0.0, 0.0, 1.0, 0.0 };
  float orientCor[] = { 1.0, 0.0, 0.0, 0.0, 0.0, 1.0 };
  float orientSag[] = { 0.0, 1.0, 0.0, 0.0, 0.0, 1.0 };
  const size_t osz = sizeof(orientTra);
  LST_HEAD *lh;
  int rval;

  typedef enum {Tra = 2, Cor = 3, Sag = 4} Orientation;

  if (!(lh = LST_Create()))
    CU_FAIL("could not create list structure");

  // checkOrientation() somewhat mysteriously requires at least three
  // entries in the parameter list, and uses only the second one.
  LST_Push(&lh, &dummy2);
  
  memcpy(&paramsTra.imageOrientation, orientTra, osz);
  LST_Push(&lh, &paramsTra);
  memcpy(&paramsCor.imageOrientation, orientCor, osz);
  LST_Push(&lh, &paramsCor);
  memcpy(&paramsSag.imageOrientation, orientSag, osz);
  LST_Push(&lh, &paramsSag);

  LST_Push(&lh, &dummy1);

  CU_ASSERT_EQUAL(Sag, checkOrientation(&lh));
  LST_Dequeue(&lh);
  CU_ASSERT_EQUAL(Cor, checkOrientation(&lh));
  LST_Dequeue(&lh);
  CU_ASSERT_EQUAL(Tra, checkOrientation(&lh));
  LST_Dequeue(&lh);

  // if there aren't at least three items in the list,
  // checkOrientation() returns an error indication.
  fprintf(stderr, "Ignore this Volume Count warning: ");
  CU_ASSERT_EQUAL(5, checkOrientation(&lh));
}

  
void test_parseParams(void) {
  int i;
  DCM_OBJECT *obj;
  DCM_ELEMENT e;
  const char *thickness = "foo thick";
  PARAMS p, p2;

  // pick numbers we're unlikely to get by accident
  const U16 valRows = 125;
  const U16 valCols = 123;
  const U16 valBitsAllocated = 5;
  const U16 valBitsStored = 3;
  const U16 valHighBit = 7;
  const U16 valPixelRep = 0;
  const char *valSeriesInstanceUID = "1.2.3.4.5.6.7.8";
  const char *valInstanceNumber = "17";
  const char *valPhotometInterp = "MONOCHROME2";
  const char *valSeriesNumber = "19";
  const char *valPixelSpacing = "1.2\\1.2";
  const char *valSliceThickness = "9.8";
  const char *valImagePosition = "1.0\\-1.0\\0.0";
  const char *valImageOrientation = "1\\0\\0\\0\\0\\1";
  const U16 valPlanarConfig = 0;
  const char *valStudyDate = "20061012";
  const char *valStudyDesc = "this was a very nice study";
  const char *valProtocol = "a great protocol";
  const char *valPatientName = "Barr^Foo^X";
  const char *valPatientID = "bf32";
  const U16 valLargestPix = 252;
  const U16 valSmallestPix = 3;
  const char *valSliceSpacing = "8.0";
  const char *valSeqName = "my sequence";
  const char *valManufacturer = "Acme";
  const char *valModelName = "BirdSeed";
  const char *valModality = "MR";
  const char *valInstitution = "WUSM";
  const char *specifiedThickness = "10.2";
  
  
  /*
   * List of tags taken from dcm_to_4dfp
   *
   * Attribute name, multiplicity, and VR individually checked against
   * DICOM standard (some disagreements with the code)
   */
  DCM_ELEMENT ELEMS_TO_PARSE[] = {
    { DCM_IMGROWS, DCM_US, "Rows", 1, sizeof(U16), (void*)&valRows },
    { DCM_IMGCOLUMNS, DCM_US, "Columns", 1, sizeof(U16), (void*)&valCols },
    { DCM_IMGBITSALLOCATED, DCM_US, "Bits Allocated", 1, sizeof(U16),
      (void*)&valBitsAllocated },
    { DCM_IMGBITSSTORED, DCM_US, "Bits Stored", 1, sizeof(U16), 
      (void*)&valBitsStored },
    { DCM_IMGHIGHBIT, DCM_US, "High Bit", 1, sizeof(U16), (void*)&valHighBit },
    { DCM_IMGPIXELREPRESENTATION, DCM_US, "Pixel Representation", 1,
      sizeof(U16), (void*)&valPixelRep },
    { DCM_RELSERIESINSTANCEUID, DCM_UI, "Series Instance UID", 1,
      DICOM_UI_LENGTH, (void*)valSeriesInstanceUID },
    { DCM_RELIMAGENUMBER, DCM_IS, "Instance Number", 1, DICOM_IS_LENGTH,
      (void*)valInstanceNumber },
    { DCM_IMGPHOTOMETRICINTERP, DCM_CS, "Photometric Interpretation", 1,
      DICOM_CS_LENGTH, (void*)valPhotometInterp },
    { DCM_RELSERIESNUMBER, DCM_IS, "Series Number", 1, DICOM_IS_LENGTH,
      (void*)valSeriesNumber},
    { DCM_IMGPIXELSPACING, DCM_DS, "Pixel Spacing", 2, DICOM_DS_LENGTH*2,
      (void*)valPixelSpacing},
    { DCM_ACQSLICETHICKNESS, DCM_DS, "Slice Thickness", 1, DICOM_DS_LENGTH,
      (void*)valSliceThickness },
    { DCM_RELIMAGEPOSITIONPATIENT, DCM_DS, "Image Position (Patient)", 3,
      DICOM_DS_LENGTH*3, (void*)valImagePosition },
    { DCM_RELIMAGEORIENTATIONPATIENT, DCM_DS, "Image Orientation (Patient)",
      6, DICOM_DS_LENGTH*6, (void*)valImageOrientation },
    { DCM_IMGPLANARCONFIGURATION, DCM_US, "Planar Configuration", 1,
      sizeof(U16), (void*)&valPlanarConfig },
    { DCM_IDSTUDYDATE, DCM_DA, "Study Date", 1, DICOM_DA_LENGTH,
      (void*)valStudyDate },
    { DCM_IDSERIESDESCR, DCM_LO, "Study Description", 1,
      DICOM_LO_LENGTH, (void*)valStudyDesc },
    { DCM_ACQPROTOCOLNAME, DCM_LO, "Protocol Name", 1, DICOM_LO_LENGTH,
      (void*)valProtocol },
    { DCM_PATNAME, DCM_PN, "Patient's Name", 1, DICOM_PN_LENGTH,
      (void*)valPatientName },
    { DCM_PATID, DCM_LO, "Patient ID", 1, DICOM_LO_LENGTH,
      (void*)valPatientID },
    { DCM_IMGLARGESTIMAGEPIXELVALUE, DCM_US, "Largest Image Pixel Value",
      1, sizeof(U16), (void*)&valLargestPix },
    { DCM_IMGSMALLESTIMAGEPIXELVALUE, DCM_US, "Smallest Image Pixel Value",
      1, sizeof(U16), (void*)&valSmallestPix },
    { DCM_ACQSLICESPACING, DCM_DS, "Spacing Between Slices", 1,
      DICOM_DS_LENGTH, (void*)valSliceSpacing },
    { DCM_ACQSEQUENCENAME, DCM_SH, "Sequence Name", 1, DICOM_SH_LENGTH,
      (void*)valSeqName },
    { DCM_IDMANUFACTURER, DCM_LO, "Manufacturer", 1, DICOM_LO_LENGTH,
      (void*)valManufacturer },
    { DCM_IDMANUFACTURERMODEL, DCM_LO, "Manufacturer's Model Name", 1,
      DICOM_LO_LENGTH, (void*)valModelName },
    { DCM_IDMODALITY, DCM_CS, "Modality", 1, DICOM_CS_LENGTH,
      (void*)valModality },
    { DCM_IDINSTITUTIONNAME, DCM_LO, "Institution Name", 1,
      DICOM_LO_LENGTH, (void*)valInstitution }
  };
  int nElems = sizeof(ELEMS_TO_PARSE)/sizeof(DCM_ELEMENT);

  if ((DCM_CreateObject(&obj, 0) != DCM_NORMAL) || !obj)
    CU_FAIL("Unable to create DCM object");

  for (i = 0; i < nElems; i++)
    if (DCM_AddElement(&obj, ELEMS_TO_PARSE+i) != DCM_NORMAL)
      CU_FAIL("Unable to add element to DCM object");

  CU_ASSERT_EQUAL(0, parseParams(&obj, 0, &p));

  CU_ASSERT_EQUAL(valRows, p.rows);
  CU_ASSERT_EQUAL(valCols, p.columns);
  CU_ASSERT_EQUAL(valBitsAllocated, p.bitsAllocated);
  CU_ASSERT_EQUAL(valBitsStored, p.bitsStored);
  CU_ASSERT_EQUAL(valHighBit, p.highBit);
  CU_ASSERT_EQUAL(valPixelRep, p.pixelRepresentation);
  CU_ASSERT_STRING_EQUAL(valSeriesInstanceUID, p.seriesUID);
  CU_ASSERT_EQUAL(atoi(valInstanceNumber), p.instanceNumber);
  CU_ASSERT_STRING_EQUAL(valPhotometInterp, p.photometricInterpretation);
  CU_ASSERT_EQUAL(atoi(valSeriesNumber), p.seriesNumber);
  CU_ASSERT_STRING_EQUAL(valPixelSpacing, p.pixelSpacing);
  CU_ASSERT_STRING_EQUAL(valSliceThickness, p.sliceThickness);
  {
    double px, py, pz;
    int n = sscanf(valImagePosition, "%lf\\%lf\\%lf", &px, &py, &pz);
    CU_ASSERT_EQUAL(3, n);
    CU_ASSERT_EQUAL(px, p.positionX);
    CU_ASSERT_EQUAL(py, p.positionY);
    CU_ASSERT_EQUAL(pz, p.positionZ);
  }
  {
    int i;
    double o[6];
    int n = sscanf(valImageOrientation, "%lf\\%lf\\%lf\\%lf\\%lf\\%lf",
		   &o[0], &o[1], &o[2], &o[3], &o[4], &o[5]);
    CU_ASSERT_EQUAL(6, n);
    for (i = 0; i < 6; i++)
      CU_ASSERT_EQUAL(o[i], p.imageOrientation[i]);
  }

  CU_ASSERT_EQUAL(valPlanarConfig, p.planarConfiguration);
  CU_ASSERT_STRING_EQUAL(valStudyDate, p.studyDate);
  CU_ASSERT_STRING_EQUAL(valStudyDesc, p.seriesDescription);
  CU_ASSERT_STRING_EQUAL(valProtocol, p.protocolName);
  CU_ASSERT_STRING_EQUAL(valPatientName, p.patientName);
  CU_ASSERT_STRING_EQUAL(valPatientID, p.patientID);

  /* These two parameters are calculated rather
   * than read from the DICOM information.
   *  CU_ASSERT_EQUAL(valLargestPix, p.imgLargestPix);
   *  CU_ASSERT_EQUAL(valSmallestPix, p.imgSmallestPix);
   */

  CU_ASSERT_STRING_EQUAL(valSliceSpacing, p.sliceSpacing);
  CU_ASSERT_STRING_EQUAL(valSeqName, p.acqSequenceName);
  CU_ASSERT_STRING_EQUAL(valManufacturer, p.idManufacturer);
  CU_ASSERT_STRING_EQUAL(valModelName, p.idModel);
  CU_ASSERT_STRING_EQUAL(valModality, p.idModality);
  CU_ASSERT_STRING_EQUAL(valInstitution, p.idInstitution);

  /* do one more with a specified slice thickness */
  CU_ASSERT_EQUAL(0, parseParams(&obj, specifiedThickness, &p2));

  /* should be all the same as p except for slice thickness */
  CU_ASSERT_STRING_EQUAL(specifiedThickness, p2.sliceThickness);
  CU_ASSERT_EQUAL(valRows, p2.rows);
  CU_ASSERT_EQUAL(valCols, p2.columns);
  CU_ASSERT_EQUAL(valBitsAllocated, p2.bitsAllocated);
  CU_ASSERT_EQUAL(valBitsStored, p2.bitsStored);
  CU_ASSERT_EQUAL(valHighBit, p2.highBit);
  CU_ASSERT_EQUAL(valPixelRep, p2.pixelRepresentation);
  CU_ASSERT_STRING_EQUAL(valSeriesInstanceUID, p2.seriesUID);
  CU_ASSERT_EQUAL(atoi(valInstanceNumber), p2.instanceNumber);
  CU_ASSERT_STRING_EQUAL(valPhotometInterp, p2.photometricInterpretation);
  CU_ASSERT_EQUAL(atoi(valSeriesNumber), p2.seriesNumber);
  CU_ASSERT_STRING_EQUAL(valPixelSpacing, p2.pixelSpacing);
  {
    double px, py, pz;
    int n = sscanf(valImagePosition, "%lf\\%lf\\%lf", &px, &py, &pz);
    CU_ASSERT_EQUAL(3, n);
    CU_ASSERT_EQUAL(px, p2.positionX);
    CU_ASSERT_EQUAL(py, p2.positionY);
    CU_ASSERT_EQUAL(pz, p2.positionZ);
  }
  {
    int i;
    double o[6];
    int n = sscanf(valImageOrientation, "%lf\\%lf\\%lf\\%lf\\%lf\\%lf",
		   &o[0], &o[1], &o[2], &o[3], &o[4], &o[5]);
    CU_ASSERT_EQUAL(6, n);
    for (i = 0; i < 6; i++)
      CU_ASSERT_EQUAL(o[i], p2.imageOrientation[i]);
  }

  CU_ASSERT_EQUAL(valPlanarConfig, p2.planarConfiguration);
  CU_ASSERT_STRING_EQUAL(valStudyDate, p2.studyDate);
  CU_ASSERT_STRING_EQUAL(valStudyDesc, p2.seriesDescription);
  CU_ASSERT_STRING_EQUAL(valProtocol, p2.protocolName);
  CU_ASSERT_STRING_EQUAL(valPatientName, p2.patientName);
  CU_ASSERT_STRING_EQUAL(valPatientID, p2.patientID);

  /* These two parameters are calculated rather
   * than read from the DICOM information.
   *  CU_ASSERT_EQUAL(valLargestPix, p2.imgLargestPix);
   *  CU_ASSERT_EQUAL(valSmallestPix, p2.imgSmallestPix);
   */

  CU_ASSERT_STRING_EQUAL(valSliceSpacing, p2.sliceSpacing);
  CU_ASSERT_STRING_EQUAL(valSeqName, p2.acqSequenceName);
  CU_ASSERT_STRING_EQUAL(valManufacturer, p2.idManufacturer);
  CU_ASSERT_STRING_EQUAL(valModelName, p2.idModel);
  CU_ASSERT_STRING_EQUAL(valModality, p2.idModality);
  CU_ASSERT_STRING_EQUAL(valInstitution, p2.idInstitution);
}

void test_testRangeText() {
  DCM_OBJECT *obj = 0;
  CONDITION rval;
  DCM_ELEMENT e;
  DCM_TAG tagIS = DCM_RELACQUISITIONNUMBER; /* VR type Integer String */
  /* should also try: DS, FL, FD, SL, SS, UL, US, maybe DA, TM, etc. */

  char buf[1024];

  if (DCM_CreateObject(&obj, 0) != DCM_NORMAL || obj == 0)
    CU_FAIL("Unable to create DCM object");

  /* we reuse this element; okay because AddElement makes deep copy */
  memset(&e, 0, sizeof(e));
  e.tag = tagIS;
  e.representation = DCM_IS;
  e.d.string = buf;
  sprintf(buf, "-5");
  e.length = sizeof(buf) - 1;

  if (DCM_AddElement(&obj, &e) != DCM_NORMAL)
    CU_FAIL("Unable to add element 1 to DCM object");

  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "-10", "-1"));
  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "-5", "-5"));
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "-10", "-8"));
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "-4", "4"));
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "-4", "+4"));
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "2", "8"));

  DCM_CloseObject(&obj);
  if (DCM_CreateObject(&obj, 0) != DCM_NORMAL || obj == 0)
    CU_FAIL("Unable to create DCM object");

  sprintf(buf, "0");
  if (DCM_AddElement(&obj, &e) != DCM_NORMAL)
    CU_FAIL("Unable to add element 1 to DCM object");
  
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "-10", "-1"));
  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "0", "0"));
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "1", "10"));
  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "-4", "+4"));

  DCM_CloseObject(&obj);
  if (DCM_CreateObject(&obj, 0) != DCM_NORMAL || obj == 0)
    CU_FAIL("Unable to create DCM object");

  sprintf(buf, "42");
  if (DCM_AddElement(&obj, &e) != DCM_NORMAL)
    CU_FAIL("Unable to add element 1 to DCM object");

  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "-45", "-40"));
  CU_ASSERT_EQUAL(1, testRangeText(&obj, tagIS, "-45", "40"));
  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "40", "45"));
  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "+40", "45"));
  CU_ASSERT_EQUAL(0, testRangeText(&obj, tagIS, "+40", "+45"));
}


/* The main() function for setting up and running the tests.
 * Returns a CUE_SUCCESS on successful running, another
 * CUnit error code on failure.
 */
int main() {
   CU_pSuite pSuite = NULL;

   /* initialize the CUnit test registry */
   if (CUE_SUCCESS != CU_initialize_registry())
      return CU_get_error();

   /* add a suite to the registry */
   pSuite = CU_add_suite("Suite_1", init_suite1, clean_suite1);
   if (NULL == pSuite) {
      CU_cleanup_registry();
      return CU_get_error();
   }

   /* add the tests to the suite */
   if (!CU_add_test(pSuite, "setBase()", test_setBase) ||
       !CU_add_test(pSuite, "testSelectionText()", test_testSelectionText) ||
       !CU_add_test(pSuite, "checkOrientation()", test_checkOrientation) ||
       !CU_add_test(pSuite, "parseParams()", test_parseParams) ||
       !CU_add_test(pSuite, "testRangeText()", test_testRangeText))
     {
       CU_cleanup_registry();
       return CU_get_error();
     }

   /* Run all tests using the CUnit Basic interface */
   CU_basic_set_mode(CU_BRM_VERBOSE);
   CU_basic_run_tests();
   CU_cleanup_registry();
   return CU_get_error();
}


flip_int.c
/*$Header: /opt/cvs/dcm_to_4dfp/flip_int.c,v 1.1 2006/08/25 20:55:49 mohanar Exp $*/

/*
**  FUNCTION: int_flip
**	
**  FUNCTIONAL DESCRIPTION: Convert integer pixel slice
                            to Analyze Coordinant System.
**  RETURN VALUE: none
**	
**  ALGORITHM: Take one slice from a image volume in 4dfp orientation
      and flip the X and Y to 4dint orientation. The Z is reordered 
      in the calling subroutine. The flag must be Analyze "hist.orient".
**	
**  SIDE EFFECTS: Assumption is that incomming data is 4dfp oriented.
**	
*/

#include "include/dicom.h"
#include <math.h>
#include <stdlib.h>
#include <stdio.h>

static void errm (char* program) {
	fprintf (stderr, "%s: memory allocation error\n", program);
	exit (-1);
}

void
int_flip(int flag, U16 *img, U16 pnx, U16 pny)
{
	U16	*vector, vecdim;
	U16	ix, iy, index;

	vecdim = pnx*pny;
	
	if (!(vector = (U16 *) malloc (vecdim * sizeof (U16)))){
	    printf(" flag=%d pnx=%d pny=%d\n",flag, pnx, pny);
	    errm ("int_flip");
        }
        
  switch(flag) {
	    
	case 0:		/* Transverse flip Y */
	    for (ix = 0; ix < pnx; ix++) {
		  for (iy = 0; iy < pny; iy++) {
			index = ix + pnx*iy;
			vector[iy] = img[index];
		  }
		  for (iy = 0; iy < pny; iy++) {
			index = ix + pnx*iy;
			img[index] = vector[pny - 1 - iy];
		  }
	    }
	    break;

	case 1:		/* Coronal flip Y */
	    for (ix = 0; ix < pnx; ix++) {
		  for (iy = 0; iy < pny; iy++) {
			index = ix + pnx*iy;
			vector[iy] = img[index];
		  }
		  for (iy = 0; iy < pny; iy++) {
			index = ix + pnx*iy;
			img[index] = vector[pny - 1 - iy];
		  }
	    }
	    break;

	case 2:		/* Sagittal: flip X and Y */
	    for (iy = 0; iy < pny; iy++) {
		  for (ix = 0; ix < pnx; ix++) {
			index = ix + pnx*iy;
			vector[ix] = img[index];
		  }
		  for (ix = 0; ix < pnx; ix++) {
			index = ix + pnx*iy;
			img[index] = vector[pnx - 1 - ix];
		  }
	    }
			/* flip Y */
	    for (ix = 0; ix < pnx; ix++) {
		  for (iy = 0; iy < pny; iy++) {
			index = ix + pnx*iy;
			vector[iy] = img[index];
		  }
		  for (iy = 0; iy < pny; iy++) {
			index = ix + pnx*iy;
			img[index] = vector[pny - 1 - iy];
		  }
	    }
	    break;
	 
	default:       /* do nothing */
	    break;
  }

} /* end of int_flip */



include/dcm_to_analyze.h
/*
          Copyright (C) 2001 Washington University

*/
/* Copyright marker.  Copyright will be inserted above.  Do not remove */
/*
**		     Electronic Radiology Laboratory
**		   Mallinckrodt Institute of Radiology
**		Washington University School of Medicine
**
** Module Name(s):	
** Author, Date:	Stephen M. Moore, 14-Oct-2001
** Intent:		
** Last Update:		$Author: mohanar $, $Date: 2006/08/25 20:55:49 $
** Source File:		$RCSfile: dcm_to_analyze.h,v $
** Revision:		$Revision: 1.1 $
** Status:		$State: Exp $
*/


typedef struct {
  void *reserved[2];
  int part10Flag;
  U16 rows;
  U16 columns;
  U16 bitsAllocated;
  U16 bitsStored;
  U16 highBit;
  U16 pixelRepresentation;
  U16 planarConfiguration;
  int instanceNumber;
  float positionX;
  float positionY;
  float positionZ;
  float imageOrientation[6];
  char seriesUID[DICOM_UI_LENGTH + 1];
  char pixelSpacing[DICOM_DS_LENGTH * 2 + 2];
  char sliceThickness[DICOM_DS_LENGTH + 1];
  char photometricInterpretation[DICOM_CS_LENGTH+1];
  char fileName[1024];
  int seriesNumber;
  
  char studyDate[DICOM_DA_LENGTH + 1];
  char seriesDescription[DICOM_PN_LENGTH + 1];
  char protocolName[DICOM_LO_LENGTH + 1];
  char patientName[DICOM_LO_LENGTH + 1];
  char patientID[DICOM_PN_LENGTH + 1];
  
  char acqSequenceName[DICOM_LO_LENGTH + 1];
  char idManufacturer[DICOM_LO_LENGTH + 1];
  char idModel[DICOM_LO_LENGTH + 1];
  char idModality[32 + 1];
  char idInstitution[DICOM_LO_LENGTH + 1];
  
  U16	imgLargestPix;
  U16	imgSmallestPix;
  
  char  sliceSpacing[DICOM_DS_LENGTH + 1];
  
} PARAMS;

void* extractPixels(DCM_OBJECT** obj, PARAMS *p);






